Blog post for ingress-nginx stabilization discussion (#28182)
kubernetes/website: content/en/blog/_posts/2021-05-30-update-with-ingress-nginx.md

---
layout: blog
title: 'Updates to Ingress Nginx with network v1'
date: 2021-05-30
slug: update-with-ingress-nginx
---

**Authors:** James Strong, Richardo Katz 

With all Kubernetes APIs, there is a process to creating, maintaining, and
ultimately deprecating them once they become GA. The network API is no
different. With the release of Kubernetes 1.22, the network API goes from

> v1beta1 to v1, with v1beta is deprecated. This move has been in discussion

Reviewer: Did you perhaps mean:
```markdown
different. With the release of Kubernetes 1.22, SomeResourceName graduates to stable.
```
?

Ingress, IngressClass and NetworkPolicy are all already GA so I'm not sure what's graduating in v1.22.

Reviewer 2: Is the entire v1beta deprecated or just v1beta1/Ingress

Reviewer: See https://kubernetes.io/docs/reference/using-api/deprecation-guide/ for details.

> The networking.k8s.io/v1beta1 API version of IngressClass will no longer be served in v1.22.
> The extensions/v1beta1 and networking.k8s.io/v1beta1 API versions of Ingress will no longer be served in v1.22. 

I think those are the relevant deprecations for the ingress-nginx code.

Author: updated using this wording.

---
layout: blog
title: 'Updates to Ingress Nginx with network v1'
date: 2021-05-30
slug: update-with-ingress-nginx
---

**Authors:** James Strong, Richardo Katz 

With all Kubernetes APIs, there is a process to creating, maintaining, and

> ultimately deprecating them once they become GA. The network API is no

Reviewer: There's a network API? Is that different from NetworkPolicy?

I know there's `networking.k8s.io/v1`; is this what we mean?

Author: Yes, networking.k8s.io/v1 . I can update the wording on that to add clarity.

Reviewer: networking.k8s.io is actually an _API group_ not an API. Ingress, within that API group, is now GA.

Also: everything in in `networking.k8s.io/v1` is GA because `v1` represents, for Kubernetes a stable API version - no `beta` or `alpha`.

Reviewer 2: My comment as well -- we should be specific. v1beta1 itself can't be deprecated because it's a just a (group, version). It's the Ingress API resource that is moving between versions.

Author: updated.

different. With the release of Kubernetes 1.22, the network API goes from
v1beta1 to v1, with v1beta is deprecated. This move has been in discussion
for quite some time now in 2017 in
[issue #43214](https://github.com/kubernetes/kubernetes/issues/43214),
[2019 with 1.16 api deprecation](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/)
, and more recently in

> [KEP-1453:Graduate Ingress API to GA.](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1453-ingress-api#122)

Reviewer: We could also hyperlink to https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22

> During our last community meeting the group has decided to continue to 
> support Kubernetes versions older than 1.22 with ingress version 0.46.0 
> for six months after the release of Kubernetes 1.22 with only CVE and 
> crashes only fixes, on a rolling basis. There will be a separate branch and 
> release of Ingress-NGINX to support this very similar to how Kubernetes 
> project proper does. As we advance the Ingress-NGINX project, the mainline 
> will support Kubernetes 1.22 and above, similar to the maintainers of the 
> Kubernetes project as a whole.

Reviewer 3: ```suggestion
During our last community meeting, the group has decided to continue supporting Kubernetes versions older than 1.22 with Ingress version 0.46.0. Support for Ingress-NGINX will continue for six months after Kubernetes 1.22 is released. Any additional bug fixes and CVEs for Ingress-NGINX will be addressed on a need-by-need basis.

Ingress-NGINX will have separate branches and releases of Ingress-NGINX to support this model, mirroring the Kubernetes project process. Future releases of the Ingress-NGINX project will track and support the latest versions of Kubernetes.
```

Reviewer 3: This was a bit difficult to read and I added in a line break and some other edits. Please let us know if any items here are not correct!

{{< table caption="Ingress NGINX supported version with Kubernetes Versions" >}}
Kubernetes Version  | Ingress-NGINX version | Support Notes
:-------------------|:----------------------|:------------
`1.22`              | TBD                   | On Going
`1.21`              | `v0.46.0`             | On Going but only CVE and crashes
`1.20`              | `v0.46.0`             | On Going but only CVE and crashes
`1.19`              | `v0.46.0`             | Support will drop 6 months after 1.22 release
{{< /table >}}    

Because of the networking v1 update going GA in Kubernetes 1.22, `v0.46.0` will 

> not work with Kubernetes 1.22. 

Reviewer 4: Should we bold? Or highlight?

Reviewer 3: I'd vote for bolding in this context instead of highlighting!

Author: bolded.

# What do I need to do

The team is currently in the process of upgrading ingress-nginx to support 
the v1 migration, you can track the progress 

> [here](https://github.com/kubernetes/ingress-nginx/pull/7156).  This also 
> means the feature requests are being frozen until the v1 update is completed. 

Reviewer 3: ```suggestion
[here](https://github.com/kubernetes/ingress-nginx/pull/7156).  Feature requests are being frozen until the v1 update is completed. 
```

In the meantime to ensure no compatibility issues: 

* Update to the latest version of Ingress-NGINX, currently [v0.46.0](https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v0.46.0) 
* After Kubernetes 1.22 is released, ensure you are using the latest version of 
  Ingress-NGINX that supports v1 networking api.

The community’s feedback and support in this effort is welcome. The
Ingress-Nginx Sub-project regularly hold community meetings where we discuss
this and other issues facing the project, meeting notes and details are
[here](https://github.com/kubernetes/community/tree/master/sig-network)

> 

Reviewer 3: ```suggestion
```
